### PR TITLE
add button to leave group chat

### PR DESCRIPTION
**Pull request by amdwit**

### CODE REVIEW

File: app/frontend/src/service/conversations.ts [hunk 1]
```diff
@@ -50,3 +51,9 @@ export async function sendMessage(groupChatId: number, text: string) {
   req.setText(text);
   return await client.conversations.sendMessage(req);
 }
+
+export async function leaveGroupChat(groupChatId: number) {
+  const req = new LeaveGroupChatReq();
+  req.setGroupChatId(groupChatId);
+  return await client.conversations.leaveGroupChat(req);
+}
```
Comment (darrenvong) [inline app/frontend/src/service/conversations.ts]:
```suggestion
export function leaveGroupChat(groupChatId: number) {
  const req = new LeaveGroupChatReq();
  req.setGroupChatId(groupChatId);
  return client.conversations.leaveGroupChat(req);
}
```
I think you can return Promise and let the function caller await the result.

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx [hunk 2]
```diff
@@ -29,6 +29,15 @@ export default function GroupChatView({
     await fetchMessages();
   };
 
+  const leaveGroupChat = async () => {
+    try {
+      await service.conversations.leaveGroupChat(groupChat.groupChatId);
+      handleClose();
+    } catch (error) {
+      setError(error.message);
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: I just realised it is possible that either of these `setX` calls (`handleClose` calls `setGroupChats` in parent tab component) could happen after the component has been unmounted from the user switching tabs while things are still loading, which is not ideal as that leads to app memory leaks as indicated from the following famous React warning I'm getting:
```
index.js:1 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
```

Comment (darrenvong) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: The best way I can think of is having to wrap these in a check that the component is still mounted before you call `setX` with a ref and `useLayoutEffect` with an empty dep that flicks the flag to true/false depending on the component's mount state.

**Review by darrenvong — CHANGES_REQUESTED** [review]

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx (hunk 2 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: We are probably not the first to encounter this. Is there some best practice described for this?

Comment (darrenvong) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: Indeed, this is a [good blog post](https://www.robinwieruch.de/react-warning-cant-call-setstate-on-an-unmounted-component) I found about it referred to by Dan in the React core team. It seems like I was on the right track - the post references class components, but the idea can be implemented with what I mentioned in a function component with hooks.

Comment (lucaslcode) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: This would work, right?

```
const [mounted, setMounted] = useState(true);
useEffect(() => () => setMounted(false), []);
```

Comment (darrenvong) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: The main problem I can see is that `useState` calls are batched so there's still a chance that a Promise resolves/rejects and tries to set state even though the component has already unmounted, whereas `mounted.current = false` from a ref would be immediate as it's setting the variable directly.

Also I believe when `useState` is initially called, the component is still initialising so it's technically not mounted until `useEffect` is called, so again a small chance that state could be set if the Promise resolves/rejects before the component is actually mounted.

Comment (darrenvong) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: So I was thinking more like this:
```typescript
function useIsMounted() {
  const isMounted = useRef(false);

  useEffect(() => {
    isMounted.current = true;
    return () => (isMounted.current = false);
  }, []);

 return isMounted.current;
}
```

Might even worth turn this into a `useIsMounted` custom hook since this will probably come up each time we need to setState in any async operations?

**Comment by amdwit** [comment]
I moved a lot of state to the model.

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx (hunk 2 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: I moved a lot of state to the model instead. Better less complexity than more.

Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: I think your example of keeping a mounted state was for me the proverbial drop that went over the bucket. The use of redux and thunks already made a big mess of the components, so I moved most of it out now.

File: app/frontend/src/features/messages/groupchats/GroupChatsTab.tsx [hunk 3]
```diff
@@ -1,72 +1,36 @@
...
-      setError(error.message);
-    }
-    setLoading(false);
-  }
-
-  function handleClose() {
-    setGroupChat(null);
-  }
+  const dispatch = useAppDispatch();
+  const state = useTypedSelector((state) => state.groupChats);
```
Comment (lucaslcode) [inline app/frontend/src/features/messages/groupchats/GroupChatsTab.tsx:16]: I'm not familiar enough with redux to know, but seems to me this could cause this component to re-render unnecessarily. Like if the list of messages changes, then the state has changed so this will re-render?

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import { createAsyncThunk } from "@reduxjs/toolkit";
+import { User } from "../../../pb/api_pb";
+import { GroupChat } from "../../../pb/conversations_pb";
+import { RootState } from "../../../reducers";
+import { service } from "../../../service";
+import { groupChatsFetched, setGroupChat, setMessages } from ".";
+
+type FetchGroupChatsArguments = void;
+
+export const fetchGroupChatsThunk = createAsyncThunk<
```
Comment (lucaslcode) [inline app/frontend/src/features/messages/groupchats/groupChatsActions.ts]: Is adding 'thunk' a good naming convention? I can see a benefit would be not forgetting to use `dispatch`

File: app/frontend/src/service/conversations.ts [hunk 5]
```diff
@@ -50,3 +51,15 @@ export async function sendMessage(groupChatId: number, text: string) {
...
+  const req = new LeaveGroupChatReq();
+  req.setGroupChatId(groupChatId);
+  return client.conversations.leaveGroupChat(req);
+}
+
+export function leaveGroupChat(groupChatId: number) {
+  const req = new LeaveGroupChatReq();
+  req.setGroupChatId(groupChatId);
+  return client.conversations.leaveGroupChat(req);
+}
```
Comment (lucaslcode) [inline app/frontend/src/service/conversations.ts]: Duplicate lines

File: app/frontend/src/stories/__mocks__/store.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import rootReducer from "../../reducers";
+import { store } from "../../store";
+
+store.replaceReducer(rootReducer);
```
Comment (lucaslcode) [inline app/frontend/src/stories/__mocks__/store.ts:4]: What does this do?

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx [hunk 7]
```diff
@@ -1,59 +1,54 @@
...
-}
+export default function GroupChatView() {
+  const dispatch = useAppDispatch();
+  const { error, loading, groupChat, messages } = useTypedSelector((state) => {
+    return state.groupChats.groupChatView;
+    // if (componentState.groupChat) {
+    //   return { ...componentState, groupChat: componentState.groupChat };
+    // } else {
+    //   throw new Error("No groupChat");
+    // }
```
Comment (lucaslcode) [inline app/frontend/src/features/messages/groupchats/GroupChatView.tsx]:
```suggestion

```

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx [hunk 8]
```diff
@@ -1,59 +1,54 @@
...
-    try {
-      const messages = await service.conversations.getGroupChatMessages(
-        groupChat.groupChatId
-      );
-      setMessages(messages);
-    } catch (error) {
-      setError(error.message);
-    }
-  }, [groupChat.groupChatId]);
+  useEffect(() => void dispatch(fetchMessagesThunk(groupChat!)), []);
```
Comment (lucaslcode) [inline app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: Why not move this to `setGroupChat` or some thunk?

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  reducers: {
+    groupChatsFetched(state, action: PayloadAction<GroupChat.AsObject[]>) {
+      state.groupChats = action.payload;
+    },
+    setGroupChat(state, action: PayloadAction<GroupChat.AsObject | null>) {
+      state.groupChatView.groupChat = action.payload;
+    },
+    setMessages(state, action: PayloadAction<Message.AsObject[]>) {
+      state.groupChatView.messages = action.payload;
+    },
```
Comment (lucaslcode) [inline app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: I think it's better to remove this and set the messages in the  `extraReducers`

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    builder
+      .addCase(fetchGroupChatsThunk.pending, (state) => {
+        state.error = "";
+        state.loading = true;
+      })
+      .addCase(fetchGroupChatsThunk.fulfilled, (state) => {
+        state.loading = false;
+      })
+      .addCase(fetchGroupChatsThunk.rejected, (state, action) => {
+        state.error = action.error.message || "";
```
Comment (lucaslcode) [inline app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: I think `action.error.message!` is better so there would be an error even though a null message shouldn't be possible.

**Review by lucaslcode — CHANGES_REQUESTED** [review]
I think since the logic is moved to a reducer, we should have some tests, whether in this PR or later.

File: app/frontend/src/features/messages/groupchats/GroupChatsTab.tsx (hunk 3 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/GroupChatsTab.tsx:16]: It will rerender when state.groupChats has changed I think. Anyway we haven't had a failing performance test yet. So no reason to optimize.

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts (hunk 4 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/groupChatsActions.ts]: I also had name doubles in some cases. But I can also check case by case.

File: app/frontend/src/stories/__mocks__/store.ts (hunk 6 above)
Comment (amdwit) [reply app/frontend/src/stories/__mocks__/store.ts:4]: originally the store came with rootReducer and persister. That persist logic is very annoying when testing. So I replaced the reducer with just the rootReducer.

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx (hunk 8 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: What do you mean? We cannot move useEffect in a thunk. We need the useEffect to load messages on opening the view.

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts (hunk 10 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: What error? From Typescript? Do you mean instead of ` || ""` ?

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts (hunk 9 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: I will check how that might look like.

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx (hunk 8 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: Ah, you mean loading the messages on setting the groupChat. Yes, that is also possible. Not sure what is best.

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts (hunk 10 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: state.error is a string. Should I change that to `string || null` ?

Comment (lucaslcode) [reply app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]:
```suggestion
        state.error = action.error.message!;
```
Like this. I assumed `|| ""` was to satisfy the typing, but if somehow the thunk is rejected but `action.error.message` is null, that would silently fail with no visible error.

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx [hunk 11]
```diff
@@ -29,13 +24,13 @@ export default function GroupChatView() {
...
-    dispatch(sendMessageThunk({ groupChat: groupChat!, text }));
+    dispatch(sendMessage({ groupChat: groupChat!, text }));
 
   const closeGroupChat = () => dispatch(setGroupChat(null));
 
-  const leaveGroupChat = () => dispatch(leaveGroupChatThunk(groupChat!));
+  const handleLeaveGroupChat = () => dispatch(leaveGroupChat(groupChat!));
 
-  useEffect(() => void dispatch(fetchMessagesThunk(groupChat!)), []);
+  useEffect(() => void dispatch(fetchMessages(groupChat!)), []);
```
Comment (lucaslcode) [inline app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: Did you try doing this in the `setGroupChat` thunk?

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts (hunk 9 above)
Comment (lucaslcode) [reply app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: [This is why I mention it](https://redux.js.org/style-guide/style-guide#model-actions-as-events-not-setters)

Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: That does not tell why it should go to the `extraReducers`. Anyway I have put it there. Can't all reducers go there?

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts (hunk 10 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: That cannot be null: https://tc39.es/ecma262/#sec-error.prototype.message if it is a normal JS error. But it is not. But I still do not understand what you want. Do you want that at runtime when action.error.message equals null that it is still put as such in state.error?

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx (hunk 8 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx]: Now it is removed.

File: app/frontend/src/features/messages/groupchats/GroupChatsTab.tsx (hunk 3 above)
Comment (darrenvong) [reply app/frontend/src/features/messages/groupchats/GroupChatsTab.tsx:16]: Isn't that the case anyway even when we were using React state? We'll probably have to look into using something like [react-window](https://github.com/bvaughn/react-window) / some form of virtual lists when we start having to render large chat histories and performance becomes an issue.

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx [hunk 12]
```diff
@@ -1,59 +1,41 @@
...
-      await fetchMessages();
-      setLoading(false);
-    })();
-  }, [fetchMessages]);
   const classes = useStyles();
   return (
     <Box className={classes.root}>
-      <Typography variant="h3">{groupChat.title}</Typography>
-      <IconButton onClick={handleClose}>
+      <Typography variant="h3">{groupChat!.title}</Typography>
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/GroupChatView.tsx:30]: I've seen in the dummy data from Aapeli's account that some group chats don't have a title. Wouldn't this crash in that case?

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+type FetchGroupChatsArguments = void;
+
+export const fetchGroupChats = createAsyncThunk<
+  void,
+  FetchGroupChatsArguments,
+  { state: RootState }
+>("hostRequests/fetchGroupChats", async (_, thunkApi) => {
+  const groupChats = await service.conversations.listGroupChats();
+  thunkApi.dispatch(groupChatsFetched(groupChats));
+});
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsActions.ts]:
```suggestion
export const fetchGroupChats = createAsyncThunk("hostRequests/fetchGroupChats", async () => {
  return service.conversations.listGroupChats();
});
```
The explicit typing on state is unnecessary as you are not retrieving state in this function, so you can simplify and let the underlying typing infer the types.

Also, since dispatching an action created from `createAsyncThunk` already auto dispatch actions for you when it resolves/rejects, it'll be simpler to return the result and set the `groupChats` state in `fetchGroupChats.fulfilled` in the slice. This is just adding extra noise for no apparent benefits and it's the kinda thing you'd do only if you were writing things by hand and when you are not using Redux toolkit.

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts (hunk 10 above)
Comment (lucaslcode) [reply app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: If it's rejected, something went wrong. If something went wrong, there should either be an error message displayed or an exception raised. As it is, if error.message is null, it looks to the user like fetchGroupChats should have worked but nothing happens.

Maybe it doesn't matter though if message always exists (I think it does for gRPC errors too)

**Review by lucaslcode — DISMISSED** [review]

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts (hunk 9 above)
Comment (lucaslcode) [reply app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]: I wasn't very clear, it's more that setting messages shouldn't be an action itself, it's part of fetching messages.

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+export const fetchMessages = createAsyncThunk<
+  void,
+  GroupChat.AsObject,
+  { state: RootState }
+>("hostRequests/fetchMessages", async (groupChat, thunkApi) => {
+  const messages = await service.conversations.getGroupChatMessages(
+    groupChat.groupChatId
+  );
+  thunkApi.dispatch(messagesFetched(messages));
+});
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsActions.ts]:
```suggestion
export const fetchMessages = createAsyncThunk("hostRequests/fetchMessages", async (groupChat) => {
  return service.conversations.getGroupChatMessages(
    groupChat.groupChatId
  );
});
```
Same as first reason as above.

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+export const setGroupChat = createAsyncThunk<
+  void,
+  GroupChat.AsObject | null,
+  { state: RootState }
+>("hostRequests/fetchMessages", async (groupChat, thunkApi) => {
+  thunkApi.dispatch(groupChatSet(groupChat));
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsActions.ts]: Why is this needed since there's no async operation here at all?

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+export const leaveGroupChat = createAsyncThunk<
+  void,
+  GroupChat.AsObject,
+  { state: RootState }
+>("hostRequests/leaveGroupChat", async (groupChat, thunkApi) => {
+  await service.conversations.leaveGroupChat(groupChat.groupChatId);
+  thunkApi.dispatch(setGroupChat(null));
+  thunkApi.dispatch(fetchGroupChats());
+});
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsActions.ts]:
```suggestion
export const leaveGroupChat = createAsyncThunk("hostRequests/leaveGroupChat", async (groupChat, thunkApi) => {
  await service.conversations.leaveGroupChat(groupChat.groupChatId);
  thunkApi.dispatch(fetchGroupChats());
});
```
Not using state here.

Group chat can also be cleared when this thunk fulfils in `leaveGroupChat.fulfilled`?

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+export const sendMessage = createAsyncThunk<
+  void,
+  { groupChat: GroupChat.AsObject; text: string },
+  { state: RootState }
+>("hostRequests/sendMessage", async ({ groupChat, text }, thunkApi) => {
+  await service.conversations.sendMessage(groupChat.groupChatId, text);
+  thunkApi.dispatch(fetchMessages(groupChat));
+});
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsActions.ts]:
```suggestion
export const sendMessage = createAsyncThunk("hostRequests/sendMessage", async ({ groupChat, text }, thunkApi) => {
  await service.conversations.sendMessage(groupChat.groupChatId, text);
  thunkApi.dispatch(fetchMessages(groupChat));
});
```
Not using state here.

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    messages: [] as Message.AsObject[],
+    error: "",
+    loading: false,
+  },
+};
+
+export type RequestsState = typeof initialState;
+export const messagesFetched = createAction<Message.AsObject[]>(
+  "messagesFetched"
+);
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsSlice.ts:26]:
```suggestion
```
This is redundant if handled in `fetchMessages.fulfilled` instead.

Also, if you need to add actions in future, it's easier to add to `reducers` in the `createSlice` helper below as that creates the action and reducer in one for you automatically behind the scenes. `createAction` should only be used when you need additional actions but you don't need a slice or it's designed to be dealt with reducers in another state slice.

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    },
+    groupChatSet(state, action: PayloadAction<GroupChat.AsObject | null>) {
+      state.groupChatView.groupChat = action.payload;
+    },
+  },
+  extraReducers: (builder) => {
+    builder
+      .addCase(messagesFetched, (state, action) => {
+        state.groupChatView.messages = action.payload;
+      })
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]:
```suggestion
```
This is redundant if handled in `fetchMessages.fulfilled` instead.

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+    groupChat.groupChatId
+  );
+  thunkApi.dispatch(messagesFetched(messages));
+});
+
+export const setGroupChat = createAsyncThunk<
+  void,
+  GroupChat.AsObject | null,
+  { state: RootState }
+>("hostRequests/fetchMessages", async (groupChat, thunkApi) => {
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsActions.ts]:
```suggestion
>("hostRequests/setGroupChat", async (groupChat, thunkApi) => {
```

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+      .addCase(fetchGroupChats.rejected, (state, action) => {
+        state.error = action.error.message!;
+        state.loading = false;
+      })
+      .addCase(fetchMessages.pending, (state) => {
+        state.groupChatView.error = "";
+        state.groupChatView.loading = true;
+      })
+      .addCase(fetchMessages.fulfilled, (state) => {
+        state.groupChatView.loading = false;
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]:
```suggestion
      .addCase(fetchMessages.fulfilled, (state, action) => {
        state.groupChatView.loading = false;
        state.groupChatView.messages = action.payload;
```

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+  if (groupChat) {
+    thunkApi.dispatch(fetchMessages(groupChat));
+  }
+});
+
+export const createGroupChat = createAsyncThunk<
+  void,
+  { title: string; users: User.AsObject[] },
+  { state: RootState }
+>("hostRequests/setGroupChat", async ({ title, users }, thunkApi) => {
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsActions.ts]:
```suggestion
>("hostRequests/createGroupChat", async ({ title, users }, thunkApi) => {
```

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+  const groupChatId = await service.conversations.createGroupChat(title, users);
+  thunkApi.dispatch(fetchGroupChats());
+  const groupChat =
+    thunkApi
+      .getState()
+      .groupChats.groupChats.find(
+        (groupChat: GroupChat.AsObject) => groupChat.groupChatId === groupChatId
+      ) || null;
+  thunkApi.dispatch(setGroupChat(groupChat));
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsActions.ts]:
```suggestion
  return thunkApi
      .getState()
      .groupChats.groupChats.find(
        (groupChat: GroupChat.AsObject) => groupChat.groupChatId === groupChatId
      ) || null;
```
So you can use it to set the `groupChat` value directly in `createGroupChat.fulfilled` instead of doing another dispatch...

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+      })
+      .addCase(fetchMessages.rejected, (state, action) => {
+        state.groupChatView.error = action.error.message || "";
+        state.groupChatView.loading = false;
+      })
+      .addCase(leaveGroupChat.pending, (state) => {
+        state.groupChatView.error = "";
+      })
+      .addCase(leaveGroupChat.fulfilled, (state) => {})
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsSlice.ts]:
```suggestion
      .addCase(leaveGroupChat.fulfilled, (state) => {
        state.groupChatView.groupChat = null;
      })
```

**Review by darrenvong — CHANGES_REQUESTED** [review]
It's better that this solves the unmounting hack, although a lot of additional complexity that can be tightened up if we use Redux toolkit more to its potential.

Also agree with @lucaslcode that tests should follow this or in a follow-up PR as this is quite a beast in terms of logic so it's quite likely I still have missed things lol

File: app/frontend/src/features/messages/groupchats/GroupChatView.tsx (hunk 12 above)
Comment (amdwit) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx:30]: It is typed as string. So no null possible.

Comment (lucaslcode) [reply app/frontend/src/features/messages/groupchats/GroupChatView.tsx:30]: It will be a blank string of there's no title

File: app/frontend/src/features/messages/groupchats/groupChatsActions.ts (hunk 13 above)
Comment (lucaslcode) [reply app/frontend/src/features/messages/groupchats/groupChatsActions.ts]: This is what I meant when I said it would be better to get rid of setMessages. I missed this last night must've been looking at the wrong diff, sorry!

**Comment by amdwit** [comment]
Thank you guys I learned a lot! 👍

File: app/frontend/src/features/messages/groupchats/groupChatsSlice.ts [hunk 25]
```diff
@@ -78,5 +90,5 @@ export const groupChatsSlice = createSlice({
   },
 });
 
-export const { groupChatsFetched, groupChatSet } = groupChatsSlice.actions;
+export const {} = groupChatsSlice.actions;
```
Comment (darrenvong) [inline app/frontend/src/features/messages/groupchats/groupChatsSlice.ts:93]:
```suggestion
```
Since we have no actions left anymore?

**Review by darrenvong — APPROVED** [review]
Only a minor non blocking change but looks good!
